File: src/main/java/cc/mirukuneko/realtrainmodrenewed/signal/SignalAspect.java
package cc.mirukuneko.realtrainmodrenewed.signal;

/**
 * RTM風の遠隔信号で使う現示一覧です。
 * ボタン順とIDを固定して、保存データとUIが同じ値を共有します。
 */
public enum SignalAspect {
    STOP(0, "停止(R)", 1),
    WARNING(1, "警戒(YY)", 4),
    CAUTION(2, "注意(Y)", 3),
    REDUCE(3, "減速(YG)", 6),
    RESTRICTED(4, "抑速", 2),
    PROCEED(5, "進行(G)", 5),
    HIGH_SPEED(6, "高速進行(GG)", 7);

    private final int id;
    private final String label;
    private final int legacyValue;

    SignalAspect(int id, String label, int legacyValue) {
        this.id = id;
        this.label = label;
        this.legacyValue = legacyValue;
    }

    public int getId() {
        return id;
    }

    public String getLabel() {
        return label;
    }

    public int getLegacyValue() {
        return legacyValue;
    }

    public static SignalAspect byId(int id) {
        for (SignalAspect aspect : values()) {
            if (aspect.id == id) {
                return aspect;
            }
        }
        return STOP;
    }

    public static SignalAspect byLegacyValue(int legacyValue) {
        for (SignalAspect aspect : values()) {
            if (aspect.legacyValue == legacyValue) {
                return aspect;
            }
        }
        return STOP;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/signal/SignalAspect.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.signal

/**
 * Signal aspects used by RTM-style remote signals.
 * Button order, IDs, and legacy values are save/UI compatibility-sensitive.
 */
enum class SignalAspect(
    @get:JvmName("getId")
    val id: Int,
    @get:JvmName("getLabel")
    val label: String,
    val translationKey: String,
    @get:JvmName("getLegacyValue")
    val legacyValue: Int,
) {
    STOP(0, "停止(R)", "signal.realtrainmodrenewed.aspect.stop", 1),
    WARNING(1, "警戒(YY)", "signal.realtrainmodrenewed.aspect.warning", 4),
    CAUTION(2, "注意(Y)", "signal.realtrainmodrenewed.aspect.caution", 3),
    REDUCE(3, "減速(YG)", "signal.realtrainmodrenewed.aspect.reduce", 6),
    RESTRICTED(4, "抑速", "signal.realtrainmodrenewed.aspect.restricted", 2),
    PROCEED(5, "進行(G)", "signal.realtrainmodrenewed.aspect.proceed", 5),
    HIGH_SPEED(6, "高速進行(GG)", "signal.realtrainmodrenewed.aspect.high_speed", 7);

    companion object {
        @JvmStatic
        fun byId(id: Int): SignalAspect {
            return entries.firstOrNull { it.id == id } ?: STOP
        }

        @JvmStatic
        fun byLegacyValue(legacyValue: Int): SignalAspect {
            return entries.firstOrNull { it.legacyValue == legacyValue } ?: STOP
        }
    }
}
